pkg/cloudprovider/rhsm/subscription_manager.go

```diff
@@ -0,0 +1,238 @@
+/*
+Copyright 2019 The Machine Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rhsm
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"k8s.io/klog"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// RedHatSubscriptionManager is responsible for removing redhat subscriptions.
+type RedHatSubscriptionManager interface {
+	UnregisterInstance(machineName string) error
+}
+
+type pagination struct {
+	Offset int `json:"offset"`
+	Limit  int `json:"limit"`
+	Count  int `json:"count"`
+}
+
+type body struct {
+	Name string `json:"name"`
+	UUID string `json:"uuid"`
+}
+
+type systemsResponse struct {
+	Pagination pagination `json:"pagination"`
+	Body       []*body    `json:"body"`
+}
+
+type credentials struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int    `json:"expires_in"`
```

Comment:
Since we have this timestamp, we can take it into account before other then `auth` API calls, since we can calculate before the call if our current access token have expired.

Reply:
Actually this is not needed and I actually added it so I do this calculation but after some scenarios testing, it simply doesn't worth to add this logic. The reason for that is, the token is renewed every 5 minutes. Thus, we try to execute a request if it passes good, if not(because the token has expired, the API will return 401 and we refresh the token.